lib/rules-generator.js

```diff
@@ -361,37 +407,49 @@ util.methods(Generator, {
       if (typeof exp.accessor == 'string') {
         accessor = exp.accessor;
       } else {
-        accessor = subExpression(exp.accessor);
+        accessor = valueExpression(subExpression(exp.accessor));
       }
-      accessor = valueExpression(accessor);
-      // snapshot references use child() wrapper - EXCEPT for built-in methods
-      // TODO: Resolve ambiguity between built-ins and user-defined property names
-      // (use functions instead)?
-      if (base.valueType == 'Snapshot' && snapshotMethods.indexOf(accessor) == -1) {
-        exp = ast.snapshotChild(base, accessor);
-      } else {
-        if (replaceExp()) {
-          exp.base = base;
-          exp.accessor = accessor;
+      // snapshot references use child() wrapper - EXCEPT for reserved methods
+      // TODO: Only do this if we are dereferencing a function call OR if
+      // there is no shadowed property in the object.
+      if (base.valueType == 'Snapshot') {
+        if (accessor == 'parent') {
+          return ast.snapshotParent(base);
+        } else if (!util.arrayIncludes(reservedMethods, accessor)) {
+          return ast.snapshotChild(base, accessor);
         }
       }
+      if (util.arrayIncludes(valueMethods, accessor)) {
+        base = valueExpression(base);
+      }
+      if (replaceExp()) {
```

Comment:
Is this an inplace modification of state? I imagine its hard to fix this now but I am sure there is a better way.